csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,313 @@
+module SummaryOutput {
+  /**
+   * Gets an output specification that specifies the qualifier in a call as
+   * the output.
+   */
+  SummaryOutput qualifier() { result = TQualifierSummaryOutput() }
+
+  /**
+   * Gets an output specification that specifies the return value from a call as
+   * the output.
+   */
+  SummaryOutput return() { result = TReturnSummaryOutput() }
+
+  /**
+   * Gets an output specification that specifies the `i`th argument in a call as
+   * the output.
+   */
+  SummaryOutput argument(int i) { result = TArgumentSummaryOutput(i) }
+
+  /**
+   * Gets an output specification that specifies parameter `j` of the delegate at
+   * argument `i` as the output.
+   */
+  SummaryOutput delegate(FrameworkCallable callable, int i, int j) {
+    result = TDelegateSummaryOutput(i, j) and
+    hasDelegateArgumentPosition2(callable, i, j)
+  }
+}
+
+class FlowSource = Impl::FlowSource;
+
+/** Provides predicates for constructing flow-source specifications. */
+module FlowSource {
+  class Range = Impl::FlowSource::Range;
+
+  private class RemoteFlowSource extends Range {
+    RemoteFlowSource() { this = "remote" }
+  }
+
+  FlowSource remote() { result = any(RemoteFlowSource s).toFlowSource() }
+
+  private class LocalFlowSource extends Range {
+    LocalFlowSource() { this = "local" }
+  }
+
+  FlowSource local() { result = any(LocalFlowSource s).toFlowSource() }
+
+  private class StoredFlowSource extends Range {
+    StoredFlowSource() { this = "stored" }
+  }
+
+  FlowSource stored() { result = any(StoredFlowSource s).toFlowSource() }
+}
+
+class FlowSink = Impl::FlowSink;
+
+/** Provides predicates for constructing flow-sink specifications. */
+module FlowSink {
+  class Range = Impl::FlowSink::Range;
+
+  private class RemoteFlowSink extends Range {
+    RemoteFlowSink() { this = "remote" }
+  }
+
+  FlowSink remote() { result = any(RemoteFlowSink s).toFlowSink() }
+
+  private class HtmlFlowSink extends Range {
+    HtmlFlowSink() { this = "html" }
+
+    override predicate isSubsetOf(FlowSink fs) { fs = remote() }
+  }
+
+  FlowSink html() { result = any(HtmlFlowSink s).toFlowSink() }
+
+  private class EmailFlowSink extends Range {
+    EmailFlowSink() { this = "email" }
+
+    override predicate isSubsetOf(FlowSink fs) { fs = remote() }
+  }
+
+  FlowSink email() { result = any(EmailFlowSink s).toFlowSink() }
+
+  private class SqlFlowSink extends Range {
+    SqlFlowSink() { this = "sql" }
+  }
+
+  FlowSink sql() { result = any(SqlFlowSink s).toFlowSink() }
+
+  private class CommandFlowSink extends Range {
+    CommandFlowSink() { this = "command" }
+  }
+
+  FlowSink command() { result = any(CommandFlowSink s).toFlowSink() }
+}
+
+class FrameworkDataFlow = Impl::FrameworkDataFlow;
+
+module Examples {
+  private import semmle.code.csharp.frameworks.system.Text
+
+  /** Data flow for `System.Text.StringBuilder`. */
+  class StringBuilderDataFlow extends FrameworkDataFlow {
```

Comment:
Should user code ever refer to this class by name? If we make this class private, then we can change its name and interface as we like in the future.

Reply:
These will be private, and are only here as examples.